Elements/src/Spatial/CellComplex.cs

```diff
@@ -0,0 +1,1285 @@
+            this.TopFaceId = topFaceId;
+        }
+    }
+
+    #endregion
+
+    /// <summary>
+    /// A geometric voxel representation for structures and potential other
+    /// </summary>
+    public class CellComplex : Elements.Element
+    {
+        [JsonIgnore]
+        private long _segmentId = 1; // we start at 1 because 0 is returned as default value from dicts
+
+        [JsonIgnore]
+        private long _directedSegmentId = 1; // we start at 1 because 0 is returned as default value from dicts
+
+        [JsonIgnore]
+        private long _vertexId = 1; // we start at 1 because 0 is returned as default value from dicts
+
+        [JsonIgnore]
+        private long _faceId = 1; // we start at 1 because 0 is returned as default value from dicts
+
+        [JsonIgnore]
+        private long _cellId = 1; // we start at 1 because 0 is returned as default value from dicts
+
+        [JsonIgnore]
+        private long _uvId = 1; // we start at 1 because 0 is returned as default value from dicts
+
+        /// <summary>
+        /// Vertices by ID
+        /// </summary>
+        public Dictionary<long, Vertex> Vertices { get; private set; } = new Dictionary<long, Vertex>();
+
+        /// <summary>
+        /// U or V directions by ID
+        /// </summary>
+        public Dictionary<long, UV> UVs { get; private set; } = new Dictionary<long, UV>();
+
+        /// <summary>
+        /// Segments by ID
+        /// </summary>
+        public Dictionary<long, Segment> Segments { get; private set; } = new Dictionary<long, Segment>();
+
+        /// <summary>
+        /// DirectedSegments by ID
+        /// </summary>
+        public Dictionary<long, DirectedSegment> DirectedSegments { get; private set; } = new Dictionary<long, DirectedSegment>();
+
+        /// <summary>
+        /// Faces by ID
+        /// </summary>
+        public Dictionary<long, Face> Faces { get; private set; } = new Dictionary<long, Face>();
+
+        /// <summary>
+        /// Cells by ID
+        /// </summary>
+        public Dictionary<long, Cell> Cells { get; private set; } = new Dictionary<long, Cell>();
+
+        /// <summary>
+        /// Tolerance for points being considered the same.
+        /// Applies individually to X, Y, and Z coordinates, not the cumulative difference!
+        /// </summary>
+        public double Tolerance = Vector3.EPSILON;
+
+        [JsonIgnore]
+        private Dictionary<double, Dictionary<double, Dictionary<double, long>>> verticesLookup = new Dictionary<double, Dictionary<double, Dictionary<double, long>>>();
+
+        [JsonIgnore]
+        private Dictionary<double, Dictionary<double, Dictionary<double, long>>> uvsLookup = new Dictionary<double, Dictionary<double, Dictionary<double, long>>>();
+
+        // Indexed by tuple of (lesserVertexId, greaterVertexId)
+        [JsonIgnore]
+        private Dictionary<(long, long), long> segmentsLookup = new Dictionary<(long, long), long>();
```

Comment:
Per our discussion of today, I think we can remove these lookups if we put the bi-directional references (vertex -> edges, edge -> faces, etc.) on the objects themselves. This will also aid in traversal.